gen3-integration-tests/utils/misc.py

```diff
@@ -0,0 +1,66 @@
+import fcntl
+import os
+import time
+
+from cdislogging import get_logger
+
+from utils import TEST_DATA_PATH_OBJECT
+
+
+logger = get_logger(__name__, log_level=os.getenv("LOG_LEVEL", "info"))
+
+
+def one_worker_only(func):
+    """
+    Use this decorator to make sure only 1 worker runs a function. Other workers will wait until the 1st
+    worker is done before proceeding. Useful for configuration autorun fixtures that must run before the
+    tests can run.
+
+    Usage:
+        @pytest.fixture(autouse=True, scope="session")
+        @one_worker_only
+        def setup_tests():
+            pass
+    """
+
+    def run_with_lock():
+        worker_id = os.environ["PYTEST_XDIST_WORKER"]
+        lock_path = TEST_DATA_PATH_OBJECT / "lock"
+
+        try:  # attempt to get a lock on the lock file
+            with open(lock_path, "w") as lock_file:  # open and create lock file
+                fcntl.flock(lock_file, fcntl.LOCK_EX | fcntl.LOCK_NB)
+                logger.info(f"Worker '{worker_id}' running '{func.__name__}'")
+
+                # if `func` completes quickly, this worker may release the lock before other workers
+                # attempt to lock, and `func` could run more than once. Wait 1 sec to avoid this
+                time.sleep(1)
+
+                result = func()
+
+                logger.info(f"Worker '{worker_id}' is done running '{func.__name__}'")
+                fcntl.flock(lock_file, fcntl.LOCK_UN)  # release lock
+                return result
+        except IOError:  # unable to lock: another worker already locked; wait
+            logger.info(
+                f"Worker '{worker_id}' waiting for '{func.__name__}' to be done running"
+            )
+            with open(lock_path, "r") as lock_file:
+                max_it = 10
```

Comment:
do we need retry logic? if one worker is running it already isn't it enough?

Reply:
this is not a retry for running the function, it's the other workers waiting until the 1st one is done running the function

Reply:
right, what happens when 2 workers are created and both try to run setup, won't it still run twice though sequentially? maybe I am misreading this

Reply:
only 1 worker runs it. here's a log from my testing earlier:
```
[2024-02-19 14:28:51,049][utils.misc][   INFO] Worker 'gw3' waiting for 'setup_tests' to be done running
[2024-02-19 14:28:51,049][utils.misc][   INFO] Worker 'gw3' waiting 3s...
[2024-02-19 14:28:54,055][utils.misc][   INFO] Worker 'gw3' is done waiting

[2024-02-19 14:28:51,049][utils.misc][   INFO] Worker 'gw2' running 'setup_tests'
[2024-02-19 14:28:52,664][utils.misc][   INFO] Worker 'gw2' is done running 'setup_tests'

[2024-02-19 14:28:51,049][utils.misc][   INFO] Worker 'gw0' waiting for 'setup_tests' to be done running
[2024-02-19 14:28:51,049][utils.misc][   INFO] Worker 'gw0' waiting 3s...
[2024-02-19 14:28:54,056][utils.misc][   INFO] Worker 'gw0' is done waiting

[2024-02-19 14:28:51,049][utils.misc][   INFO] Worker 'gw1' waiting for 'setup_tests' to be done running
[2024-02-19 14:28:51,049][utils.misc][   INFO] Worker 'gw1' waiting 3s...
[2024-02-19 14:28:54,055][utils.misc][   INFO] Worker 'gw1' is done waiting
```